src/persist/src/indexed/mod.rs

```diff
@@ -689,24 +726,78 @@ impl<L: Log, B: Blob> Indexed<L, B> {
             self.blob.delete_unsealed_batch(&batch)?;
         }
 
-        for batch in deleted_trace_batches {
-            self.blob.delete_trace_batch(&batch)?;
+        Ok(())
+    }
+
+    fn compact_trace_external_reqs(&mut self) -> Result<Vec<ExternalWorkReq>, Error> {
+        let mut reqs = vec![];
+        for (id, arrangement) in self.state.arrangements.iter() {
+            if self.in_flight_trace_compactions.contains(&id) {
+                continue;
+            }
+            if let Some(req) = arrangement.trace_next_compact_req(&self.maintainer)? {
+                reqs.push(ExternalWorkReq::CompactTrace(req));
+                self.in_flight_trace_compactions.insert(*id);
+            }
         }
+        Ok(reqs)
+    }
+
+    fn do_handle_external_work_res(&mut self, res: ExternalWorkRes) -> Result<(), Error> {
+        match res {
+            ExternalWorkRes::CompactTrace(res) => {
+                match res {
+                    Ok(res) => {
+                        let written_bytes = res.merged.size_bytes;
+                        self.metrics.compaction_write_bytes.inc_by(written_bytes);
+                        self.apply_batched_cmd(|state, pending| {
+                            let id = res.req.id;
+                            if !state.arrangements.contains_key(&id) {
+                                log::trace!(
+                                    "trace compaction response for deleted trace id: {:?}",
+                                    id
+                                );
+                                return;
+                            }
+                            let arrangement = state
+                                .arrangements
+                                .get_mut(&id)
+                                .expect("arrangement known to exist");
+                            let (_, deleted_trace_batches) =
+                                arrangement.trace_handle_compact_response(res);
+
+                            pending.add_deleted_trace_batches(id, deleted_trace_batches);
+                        });
+                    }
+                    Err(e) => {
+                        // TODO: flesh this out more
+                        log::trace!("error in external work result: {:?}", e);
+                    }
+                }
+            }
+        };
 
         Ok(())
     }
 
+    /// Handle the results of a previously sent external work request.
+    pub fn handle_external_work_res(&mut self, res: ExternalWorkRes) {
+        if let Err(e) = self.do_handle_external_work_res(res) {
+            log::trace!("error handling external work result: {:?}", e);
```

Comment:
is there something better to do when handling an external work response error? this is how we handle `step` errors so it doesn't seem like a huge change, but feels gross.

Reply:
done.